website/blog/2023-12-15-serverless-free-tier-data-stack-with-dlt-and-dbt-core.md

```diff
@@ -0,0 +1,168 @@
+
+**The builder:** I’m a data freelancer who deploys end to end solutions, so when I have a data problem, I cannot just let it go. 
+
+**The tools:** I want to be able to run my project on `google cloud functions` due to the generous free tier. I will try `dlt` for ingestion, which I wanted to test for some time. And for transformation I will use `dbt core`.
+
+## The starting point
+
+My partner and I are considering buying a property in Portugal. Unfortunately, the only data source we have is Idealista, which is a portal where real estate agencies post ads.
+
+Unfortunately, there are way fewer properties than ads - it seems many real estate companies re-post the same ad that others do, with intentionally different data and often misleading bits of info. The real estate agencies do this so the interested parties reach out to them for clarification, and from there they can start a sales process. At the same time, the website with the ads is incentivised to allow this to continue as they get paid per ad, not per property.
+
+So it seems if I want to collect some information I will need to 
+
+- Grab the data and historize it
+- Deduplicate existing listings
+- Try to infer what listings sold for how much.
+
+Once we have deduplicated listings with some online history, we can get an idea 
+
+- How expensive which properties are
+- How fast they get sold, hopefully a signal of whether they are “worth it” or not.
+
+## Towards a solution
+
+The solution has pretty standard components
+
+- An EtL pipeline. The little t stands for normalisation, such as transforming strings to dates or unpacking nested structures.
+
+Due to the complexity of deduplication, we needed to add a human element to confirm the deduplication. This is reflected in the diagram below:
+
+<Lightbox src="/img/blog/serverless-free-tier-data-stack-with-dlt-and-dbt-core/architecture_diagram.png" width="70%" title="Project architecture" />
+
+### Ingesting the data
+
+For ingestion, we use a couple of sources
+
+- A freemium api wrapper for Idealista found at: https://rapidapi.com/apidojo/api/idealista2
+- A google sheet for manually confirming the deduplication. To make this work, we first load the data to BigQuery then access it via the sheet client. We then do manual annotation before re-loading this data back to BigQuery. For this we will use the [ready-made dlt sheets source connector](https://dlthub.com/docs/dlt-ecosystem/verified-sources/google_sheets).
+
+Since these steps run sequentially, I chose to separate them into their own repositories:
+
+- Idealista pipeline: https://github.com/euanjohnston-dev/Idealista_pipeline
+- gsheets deduplication pipeline https://github.com/euanjohnston-dev/gsheets_check_pipeline
```

Comment:
```suggestion
First, I ingest home listings from the Idealista API, accessed through [API Dojo's freemium wrapper](https://rapidapi.com/apidojo/api/idealista2). The dlt pipeline I created for ingestion is in [this repo](https://github.com/euanjohnston-dev/Idealista_pipeline).

After an initial round of transformation (described in the next section), the deduplicated data is loaded into BigQuery where I can query it from the Google Sheets client and manually review the deduplication. 

When I'm happy with the results, I use the [ready-made dlt Sheets source connector](https://dlthub.com/docs/dlt-ecosystem/verified-sources/google_sheets) to pull the data back into BigQuery, [as defined here](https://github.com/euanjohnston-dev/gsheets_check_pipeline).
```

Reworked this a bit to remove the naked links and make the Google Sheets section clearer - we're in the ingestion section, but you were talking about exposing data to GSheets for manual annotation.